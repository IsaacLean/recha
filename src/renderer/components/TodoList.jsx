// @flow
import React from 'react'

import Todo from './Todo'
import type { Todo as TodoType } from '../types'

const TodoList = ({ onTodoClick, todos }: { onTodoClick?: (...args: Array<any>) => any, todos: Array<TodoType> }) => (
  <ul>
    {todos.map(todo => {
      const props = {
        ...todo,
        onClick: undefined
      }

      if(onTodoClick) {
        props.onClick = () => onTodoClick(todo.id)
      }

      return <Todo key={todo.id} {...props} />
    })}
  </ul>
)

export default TodoList
