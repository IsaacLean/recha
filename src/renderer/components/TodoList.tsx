import React from 'react'

import Todo from './Todo'
import { Todo as TodoType } from '../types'

const TodoList = ({ onTodoClick, todos }: { onTodoClick?: (...args: any[]) => any, todos: TodoType[] }): JSX.Element => (
  <ul>
    {todos.map(todo => {
      const props = {
        ...todo,
        onClick: undefined
      }

      if(onTodoClick) {
        props.onClick = () => onTodoClick(todo.id)
      }

      return <Todo key={todo.id} {...props} />
    })}
  </ul>
)

export default TodoList
