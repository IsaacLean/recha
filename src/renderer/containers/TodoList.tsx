import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DragDropContext, Droppable } from 'react-beautiful-dnd'

import AddTodo from '../components/AddTodo'
import Todo from './Todo'
import { default as TodoType } from '../types/Todo'
import { fetchTodos as acFetchTodos } from '../actions/todos'
import { getTodoList } from '../reducers/todos'
import { NormalizedTodosRes } from '../types/actions'
import { State as ReduxState } from '../types/reducers'

interface DispatchProps {
  fetchTodos: (date: string) => Promise<NormalizedTodosRes>
}

interface StateProps {
  date: string
  reduxTodoList: TodoType[]
}

interface InnerListProps {
  todoList: TodoType[]
}

type Props = StateProps & DispatchProps

interface ComponentState {
  compTodoList: TodoType[]
}

class InnerList extends Component<InnerListProps> {
  public shouldComponentUpdate(nextProps): boolean {
    const { todoList } = this.props
    if (nextProps.todoList === todoList) {
      return false
    }
    return true
  }

  public render(): JSX.Element[] {
    const { todoList } = this.props
    return todoList.map((todo, index) => <Todo key={todo.id} todo={todo} index={index} />)
  }
}

// eslint-disable-next-line react/no-multi-comp
class TodoList extends Component<Props, ComponentState> {
  public state = {
    compTodoList: []
  }

  public componentDidMount(): void {
    const { fetchTodos } = this.props
    fetchTodos('2019-02-03')
  }

  public componentDidUpdate(prevProps): void {
    const { reduxTodoList } = this.props
    if (prevProps.reduxTodoList !== reduxTodoList) {
      this.setState({ compTodoList: reduxTodoList }) // eslint-disable-line react/no-did-update-set-state
    }
  }

  private applyStyles = isDraggingOver => {
    let styles = ''
    if (isDraggingOver) {
      styles += 'todo-list-drag-over'
      return styles
    }
    return styles
  }

  private onDragEnd = result => {
    const { compTodoList } = this.state
    const { destination, source } = result

    if (!destination) return
    if (destination.droppableId === source.droppableId && destination.index === source.index) return

    const newUITodoList = Array.from(compTodoList)
    const [movedTodo] = newUITodoList.splice(source.index, 1)
    newUITodoList.splice(destination.index, 0, movedTodo)

    this.setState({ compTodoList: newUITodoList })
  }

  public render(): JSX.Element {
    const { date, fetchTodos } = this.props
    const { compTodoList } = this.state

    let dndList
    if (date) {
      dndList = (
        <DragDropContext onDragEnd={this.onDragEnd}>
          <h1>{date}</h1>
          <AddTodo />
          <Droppable droppableId={date}>
            {(provided, snapshot) => (
              <div
                ref={provided.innerRef}
                {...provided.droppableProps}
                className={this.applyStyles(snapshot.isDraggingOver)}
              >
                <InnerList todoList={compTodoList} />
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </DragDropContext>
      )
    } else {
      dndList = 'Empty List'
    }

    return dndList
  }
}

const mapStateToProps = (state: ReduxState): StateProps => ({
  date: state.todos.list.id as string,
  reduxTodoList: getTodoList(state.todos)
})

const mapDispatchToProps = {
  fetchTodos: acFetchTodos
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoList)
