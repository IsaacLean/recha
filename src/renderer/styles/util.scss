.util-cursor-grab {
  cursor: grab !important;
}

.util-cursor-pointer {
  cursor: pointer !important;
}

.util-display-block {
  display: block !important;
}

.util-display-none {
  display: none !important;
}
