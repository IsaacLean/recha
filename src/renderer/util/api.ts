import fetch from 'cross-fetch'

import HTTPErr from './HTTPErr'
import Todo from '../types/Todo'
import { APIRes } from '../types/index2'

const rootPath = '/api/v1'
// if(process.env.NODE_ENV === 'test') {
//   rootPath = `http://localhost:9001${rootPath}`
// }

interface TodosQuery {
  col: 'id' | 'date' | 'order_num'
  date: string
  dir: 'ASC' | 'asc' | 'DESC' | 'desc'
}

// eslint-disable-next-line import/prefer-default-export
export const getTodos = (query?: Partial<TodosQuery>): Promise<APIRes<Todo[]>> => {
  let path = `${rootPath}/todos`

  if (query) {
    const queryString = new URLSearchParams(query).toString()
    if (queryString) path += `?${queryString}`
  }

  return fetch(path).then(res => {
    if (!res.ok) throw new HTTPErr(res.statusText, res)
    return res.json()
  })
}
