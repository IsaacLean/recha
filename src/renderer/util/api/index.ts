import * as todos from './todos'

type Method = 'DELETE' | 'GET' | 'PATCH' | 'POST' | 'PUT'

interface ReqOptions {
  method: string
  headers: { 'Content-Type': 'application/json' }
  body?: any // eslint-disable-line @typescript-eslint/no-explicit-any
}

export const ROOT_PATH = '/api/v1'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const genReqOptions = (method: Method, data?: any): ReqOptions => {
  const reqOptions: ReqOptions = {
    method,
    headers: { 'Content-Type': 'application/json' }
  }
  if (data) reqOptions.body = JSON.stringify(data)
  return reqOptions
}

export { todos }
