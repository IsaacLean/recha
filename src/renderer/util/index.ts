/* eslint-disable import/prefer-default-export */
export const todoIsChecked = (completed_at: string): boolean => {
  if (completed_at) return true
  return false
}
