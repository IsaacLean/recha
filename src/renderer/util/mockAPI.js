import { v4 } from 'uuid'

const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

export const mockDatabase = {
  todos: [{
    id: '24236ed7-e956-4cad-90e9-65469c554fc0',
    text: 'hey',
    completed: true
  }, {
    id: '3df29b26-8f55-4722-abbd-428a3b3d3306',
    text: 'ho',
    completed: true
  }, {
    id: 'afe441fd-b058-4eef-be6c-867f9956eba3',
    text: 'let us go',
    completed: false
  }]
}

export const addTodo = text =>
  delay(500).then(() => {
    const todo = {
      id: v4(),
      text,
      completed: false
    }
    mockDatabase.todos.push(todo)
    return todo
  })

export const fetchTodos = filter => delay(500).then(() => {
  // if(Math.random() > 0.5) {
  //   throw new Error('Boom!')
  // }

  switch(filter) {
    case 'all':
      return mockDatabase.todos
    case 'active':
      return mockDatabase.todos.filter(t => !t.completed)
    case 'completed':
      return mockDatabase.todos.filter(t => t.completed)
    default:
      throw new Error(`Unknown filter: ${filter}`)
  }
})

export const toggleTodo = id =>
  delay(500).then(() => {
    const todo = mockDatabase.todos.find(t => t.id === id)
    todo.completed = !todo.completed
    return todo
  })
