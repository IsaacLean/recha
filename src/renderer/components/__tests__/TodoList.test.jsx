import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'

import TodoList from '../TodoList'
import { mockDatabase } from '../../util/mockAPI'

Enzyme.configure({ adapter: new Adapter() })

describe('TodoList', () => {
  it('renders properly', () => {
    const component = shallow(<TodoList todos={mockDatabase.todos} />)
    expect(component).toMatchSnapshot()
  })
})
