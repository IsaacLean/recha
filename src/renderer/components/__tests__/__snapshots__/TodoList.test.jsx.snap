// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders properly 1`] = `
<ul>
  <Todo
    completed={true}
    id="24236ed7-e956-4cad-90e9-65469c554fc0"
    key="24236ed7-e956-4cad-90e9-65469c554fc0"
    text="hey"
  />
  <Todo
    completed={true}
    id="3df29b26-8f55-4722-abbd-428a3b3d3306"
    key="3df29b26-8f55-4722-abbd-428a3b3d3306"
    text="ho"
  />
  <Todo
    completed={false}
    id="afe441fd-b058-4eef-be6c-867f9956eba3"
    key="afe441fd-b058-4eef-be6c-867f9956eba3"
    text="let us go"
  />
</ul>
`;
